slatedb/src/batch_write.rs

```diff
@@ -91,7 +91,18 @@ impl DbInner {
         // get the durable watcher. we'll await on current WAL table to be flushed if wal is enabled.
         // otherwise, we'll use the memtable's durable watcher.
         let durable_watcher = if self.wal_enabled {
-            let current_wal = self.wal_buffer.maybe_trigger_flush().await?;
+            let current_wal = match self.wal_buffer.maybe_trigger_flush().await {
+                Ok(current_wal) => current_wal,
+                Err(err) => {
+                    // maybe_trigger_flush returns an error when the WAL buffer is in a fatal error state
+                    // (excluding BackgroundTaskShutdown). This indicates the flushing error is not
+                    // recoverable, we should `record_fatal_error` to notify others.
+                    if !matches!(err, SlateDBError::BackgroundTaskShutdown) {
```

Comment:
IIRC, we set fatal state even for background task shutdown. See the `cleanup_fn`'s for examples. So I think we should always set the fatal state to err.clone().

Reply:
I also wonder if we should leave this code as it is, and instead raise the error in `spawn_write_task`:

```rust
                    WriteBatchMsg::WriteBatch(write_batch_request) => {
                        let WriteBatchRequest { batch, done } = write_batch_request;
                        let result = this.write_batch(batch).await;
                        _ = done.send(result);
                        // TODO raise an error if result is Err
                    }
```

Then, the `cleanup_fn` will set the fatal state (it already has that code).